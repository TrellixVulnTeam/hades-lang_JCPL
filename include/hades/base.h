#ifndef HADES_BASE_H
#define HADES_BASE_H

#include "hades/base/sys.h"
#include "hades/base/data.h"

namespace hades {

#define HADES_DEFAULT_COPY(T) T(const T&) = default; \
  auto operator=(const T&) -> T& = default;
#define HADES_DEFAULT_MOVE(T) T(T&&) = default; \
  auto operator=(T&&) -> T& = default;

#define HADES_DELETE_COPY(T) T(const T&) = delete; \
  auto operator=(const T&) -> T& = delete;

template <typename T>
constexpr bool is_copyable_v = std::is_copy_constructible_v<T> && std::is_copy_assignable_v<T>;
template <typename T>
constexpr bool is_movable_v = std::is_move_constructible_v<T> && std::is_move_assignable_v<T>;

} // namespace hades


#endif