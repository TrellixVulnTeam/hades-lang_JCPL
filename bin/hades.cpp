#include "llvm/IR/LLVMContext.h"
#include "hades/core/Context.h"

auto main() -> int {
    hades::core::Context ctx;
    ctx.run();
    return 0;
}