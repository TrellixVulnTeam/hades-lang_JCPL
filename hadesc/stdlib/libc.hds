
extern def puts(*Byte): Void = puts;

