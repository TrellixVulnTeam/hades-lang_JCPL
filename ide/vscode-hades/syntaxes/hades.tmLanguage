<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>$schema</key>
    <string>https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json</string>
    <key>name</key>
    <string>Hades</string>
    <key>patterns</key>
    <array>
      <dict>
        <key>include</key>
        <string>#declarations</string>
      </dict>
    </array>
    <key>fileTypes</key>
    <array>
      <string>hds</string>
    </array>
    <key>repository</key>
    <dict>
      <key>declarations</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#extern-function-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#function-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#import-as</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#struct-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#const-def</string>
          </dict>
        </array>
      </dict>
      <key>extern-function-def</key>
      <dict>
        <key>begin</key>
        <string>\b(extern)\s+</string>
        <key>end</key>
        <string>;</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>storage.modifier.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-signature</string>
          </dict>
        </array>
      </dict>
      <key>function-signature</key>
      <dict>
        <key>begin</key>
        <string>\b(def)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>[=;}]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function.hades</string>
          </dict>
        </dict>
      </dict>
      <key>function-def</key>
      <dict>
        <key>begin</key>
        <string>\b(def)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-def-signature</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
        </array>
      </dict>
      <key>function-def-signature</key>
      <dict>
        <key>begin</key>
        <string>(?&lt;=(def)\s+([a-zA=Z_][a-zA-Z0-9_]*))</string>
        <key>end</key>
        <string>(?=[{])</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-def-signature-param-list</string>
          </dict>
        </array>
      </dict>
      <key>function-def-signature-param-list</key>
      <dict>
        <key>begin</key>
        <string>(\()</string>
        <key>end</key>
        <string>\)</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#param</string>
          </dict>
        </array>
      </dict>
      <key>param</key>
      <dict>
        <key>begin</key>
        <string>\b([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?=[,)])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>variable.parameter</string>
          </dict>
        </dict>
      </dict>
      <key>import-as</key>
      <dict>
        <key>begin</key>
        <string>\b(import)\s+([a-zA-Z_][a-zA-Z0-9]*)\b(\.[a-zA-Z_][a-zA-Z0-9_]*)*\s+(as)\s+([a-zA-Z_][a-zA-Z0-9_]*)</string>
        <key>end</key>
        <string>[;
]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.class.hades</string>
          </dict>
          <key>3</key>
          <dict>
            <key>name</key>
            <string>entity.name.class.hades</string>
          </dict>
          <key>4</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>5</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
      </dict>
      <key>struct-def</key>
      <dict>
        <key>begin</key>
        <string>\b(struct)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.class</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#keywords</string>
          </dict>
        </array>
      </dict>
      <key>block</key>
      <dict>
        <key>begin</key>
        <string>([{])</string>
        <key>end</key>
        <string>([}])</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#statements</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#if-statement</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>statements</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#val-statement</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#return-statement</string>
          </dict>
        </array>
      </dict>
      <key>val-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(val)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>([;
])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>return-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(return)\b</string>
        <key>end</key>
        <string>([;
])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.control</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>if-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(if)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.control.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>expressions</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#boolean-literals</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#string-literal</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-expressions</string>
          </dict>
        </array>
      </dict>
      <key>boolean-literals</key>
      <dict>
        <key>match</key>
        <string>\b(true|false)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>constant.language</string>
          </dict>
        </dict>
      </dict>
      <key>string-literal</key>
      <dict>
        <key>begin</key>
        <string>((b?)(["]))</string>
        <key>end</key>
        <string>(["])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>emphasis</string>
          </dict>
          <key>3</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double</string>
          </dict>
        </dict>
        <key>captures</key>
        <dict>
          <key>0</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#inside-string</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#string-escapes</string>
          </dict>
        </array>
      </dict>
      <key>inside-string</key>
      <dict>
        <key>match</key>
        <string>([^"\\]+)</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double.hades</string>
          </dict>
        </dict>
      </dict>
      <key>string-escapes</key>
      <dict>
        <key>match</key>
        <string>(\\[n\\t"])</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>constant.character.escape</string>
          </dict>
        </dict>
      </dict>
      <key>keywords</key>
      <dict>
        <key>match</key>
        <string>\b(val)\b</string>
        <key>name</key>
        <string>keyword</string>
      </dict>
      <key>const-def</key>
      <dict>
        <key>begin</key>
        <string>\b(const)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>[;
]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>variable</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>variable-expressions</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#variable-method</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-call</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-lowercase</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-uppercase</string>
          </dict>
        </array>
      </dict>
      <key>variable-lowercase</key>
      <dict>
        <key>match</key>
        <string>\b([a-z_][a-zA-Z0-9_]*)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
      </dict>
      <key>variable-uppercase</key>
      <dict>
        <key>match</key>
        <string>\b([A-Z_][a-zA-Z0-9_]*)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>entity.name.class</string>
          </dict>
        </dict>
      </dict>
      <key>variable-call</key>
      <dict>
        <key>match</key>
        <string>\b([a-zA-Z_][a-zA-Z0-9_]*)\(\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>entity.name.function</string>
          </dict>
        </dict>
      </dict>
      <key>variable-method</key>
      <dict>
        <key>match</key>
        <string>\b(\.)(([a-zA-Z_][a-zA-Z0-9_]*))\(\b</string>
        <key>captures</key>
        <dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function</string>
          </dict>
        </dict>
      </dict>
    </dict>
    <key>scopeName</key>
    <string>hades</string>
  </dict>
</plist>