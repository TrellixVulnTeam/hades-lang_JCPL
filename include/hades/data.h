#ifndef HADES_DATA_H
#define HADES_DATA_H

#include "memory"
#include "string"
#include "string_view"
#include "vector"
#include "llvm/ADT/SmallVector.h"

namespace hades {
using String = std::string;
using StringView = std::string_view;
using StringRef = llvm::StringRef;
using Twine = llvm::Twine;

template <typename T, typename Alloc = std::allocator<T>>
using Vec = std::vector<T, Alloc>;

template <typename T, unsigned Size>
using SmallVec = llvm::SmallVector<T, Size>;
} // namespace hades

#endif
