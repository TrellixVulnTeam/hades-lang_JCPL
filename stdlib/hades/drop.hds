
interface Drop {
    def drop(this: *mut This): Void;
}

