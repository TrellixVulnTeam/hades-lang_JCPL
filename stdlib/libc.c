#include <dirent.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>

const char* hdc_dirent_name(struct dirent* entry) {
    return entry->d_name;
}

unsigned char hdc_dirent_type(struct dirent* entry)  {
    return entry->d_type;
}

unsigned char hdc_dirent_is_directory(struct dirent* entry) {
    return S_ISDIR(entry->d_type);
}

void hdc_set_errno(int value) {
    errno = value;
}


FILE* hdc_get_stdout() {
    return stdout;
}


int hdc_errno() {
    return errno;
}