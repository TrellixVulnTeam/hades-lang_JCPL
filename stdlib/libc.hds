
extern def puts(*Byte): Void = puts;
extern def exit(CInt): Void = exit;
extern def malloc(Size): *mut Void = __hds_malloc;
extern def free(*mut Void): Void = __hds_free;
extern def strlen(*Byte): Size = strlen;
extern def memcpy(*Byte, *Byte, Size): Void = memcpy;
extern def strlen(*Byte): Size = strlen;
/// we don't have variadics right now
/// there isn't a better way to type this
extern def fprintf(*File, *Byte, *Byte): Size = fprintf;
extern def printf(*Byte, *Byte): Size = printf;

struct File {}

