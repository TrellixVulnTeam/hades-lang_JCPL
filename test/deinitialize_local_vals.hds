import libc as c;
import hades.drop as drop;
struct X {
    val mut message: *Byte;
}

implement drop.Drop for X {
    def drop(this: *mut X): Void {
        c.puts(*this.message);
    }
}

def main(): Void {
    val x1 = X(b"x1");
    val x2 = X(b"x2");
    val mut x3 = x2;
    x3.message = b"x3";

    f();
    c.puts(b"f returned");

    if true {
        val x5 = X(b"x5");
    } else {
        val x5 = X(b"x5_unreachable");
    }
    c.puts(b"branch end");
    c.puts(b"returning from main");
}

def f(): Void {
    c.puts(b"f called");
    val x4 = X(b"x4");
    c.puts(b"returning from f");
}