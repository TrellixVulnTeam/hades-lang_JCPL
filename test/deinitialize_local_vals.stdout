f called
returning from f
x4
f returned
x5
branch end
returning from main
x3
x2
x1